# Implementação de Rede Neural Perceptron

Renato Barros

## Introdução

No cenário dinâmico do mundo corporativo, a capacidade de tomar decisões
inteligentes e estratégicas é fundamental para o sucesso. Neste
contexto, o uso do Perceptron, um algoritmo de aprendizado de máquina
inspirado no funcionamento do cérebro humano, pode ser uma ferramenta
poderosa para análise de dados e previsão de comportamento.

O Perceptron é como um aprendiz dedicado que, ao observar padrões e
relações em dados, aprende a fazer previsões e auxiliar na tomada de
decisões. Seja na identificação de produtos com potencial de venda, na
detecção de atividades fraudulentas ou na personalização da experiência
do cliente, ele oferece uma gama de aplicações para impulsionar a
eficiência e a inteligência das empresas atuais.

## Funcionamento

O Perceptron funciona como uma rede neural artificial básica: recebe
informações, processa-as e gera um resultado. Esse processamento envolve
alguns cálculos matemáticos simples:

1.  **Combinação Linear:** Cada informação (representada por um valor
    numérico) é multiplicada por um peso, que indica sua importância na
    decisão final. Os resultados dessas multiplicações são somados,
    gerando um valor intermediário chamado de “soma ponderada”.

    Fórmula: `u = (x1 * peso1) + (x2 * peso2) + ...`

2.  **Função de Ativação:** A “soma ponderada” (u), calculada na etapa
    anterior, é então passada por uma função de ativação, que define o
    resultado final do Perceptron. A função de ativação introduz não
    linearidade ao modelo, permitindo que ele aprenda relações complexas
    entre os dados. No nosso caso, usaremos a função “degrau” (step
    function), que funciona como um interruptor:

    Fórmula:

    `f(u) =       1, se u >= 0      0, se u < 0`

    Onde:

    -   `f(u)` é a saída da função de ativação.
    -   `u` é a “soma ponderada” calculada na etapa anterior.

    Em outras palavras, se a “soma ponderada” for maior ou igual a zero,
    a saída do Perceptron é 1; caso contrário, a saída é 0. Essa saída
    binária pode representar, por exemplo, a recomendação ou não de um
    produto, a classificação de um cliente como potencial comprador ou
    qualquer outro evento que se deseje prever.

3.  **Função Delta:** Após calcular a saída usando a função de ativação,
    o Perceptron compara essa saída com a saída desejada (o valor
    correto que ele deveria ter previsto). A diferença entre a saída
    desejada e a saída real é chamada de “erro”. Para minimizar esse
    erro e “aprender” com seus erros, o Perceptron ajusta seus pesos
    usando a regra delta (ou regra de Widrow-Hoff). Essa regra ajusta
    cada peso proporcionalmente ao erro e à entrada correspondente.

    Fórmula:
    `novo_peso = peso_atual + (taxa_de_aprendizagem * erro * entrada)`

    Onde:

    -   `taxa_de_aprendizagem` é um valor pequeno que controla a
        velocidade do aprendizado (como visto na seção “Parâmetros e
        Dados de Treinamento”).

Para ilustrar o funcionamento do Perceptron na prática, temos este
código Python que implementa um modelo simples, utilizando exemplos e
analogias do universo corporativo.

## Parâmetros e Dados de Treinamento

Assim como um estagiário precisa de orientação inicial, o Perceptron
também precisa de parâmetros iniciais e dados para iniciar seu
aprendizado.

In [1]:
# Implementação de Rede Neural Perceptron
# Copyright (C) 2026 Renato Barros
#
# Este programa é um software livre: você pode redistribuí-lo e/ou modificá-lo
# sob os termos da Licença Pública Geral GNU (GNU General Public License)
# conforme publicada pela Free Software Foundation, seja a versão 3 da Licença
# ou (a seu critério) qualquer versão posterior.
#
# Este programa é distribuído na esperança de que seja útil, mas SEM QUALQUER
# GARANTIA; sem mesmo a garantia implícita de COMERCIABILIDADE ou ADEQUAÇÃO A
# UM DETERMINADO PROPÓSITO. Veja a Licença Pública Geral GNU para mais detalhes.
#
# Você deve ter recebido uma cópia da Licença Pública Geral GNU junto com
# este programa. Se não, veja <https://www.gnu.org/licenses/>.

# Inicializa os pesos do perceptron (w1 e w2)
peso = [0.2, -0.1]
# Define a taxa de aprendizagem para o algoritmo
taxa_aprendizagem = 0.1
# Inicializa o contador de erros totais do perceptron como zero
erros_totais = 0.0
# Define o limite de épocas (iterações) para o treinamento
limite_epocas = 20

# Define os dados de treinamento com seus respectivos valores
# de entrada (x1 e x2) e a saída desejada (valor de referência)
dados_treinamento = [
    {"x1": 0.5, "x2": 0.8, "saida_desejada": 1},
    {"x1": 0.2, "x2": 0.4, "saida_desejada": 0},
]

-   **Inicializando os Pesos:** No Perceptron, os “pesos” representam a
    importância de cada informação na tomada de decisão. No código,
    `peso = [0.2, -0.1]` significa que estamos atribuindo pesos iniciais
    aleatórios para duas informações relevantes (x1 e x2).

    Digamos, por exemplo, que estamos construindo um modelo para prever
    a probabilidade de um produto ser um sucesso de vendas. Nesse caso,
    x1 poderia ser o número de avaliações positivas do produto e x2 a
    quantidade de vezes que ele foi adicionado à lista de desejos dos
    clientes. Os pesos (0.2 e -0.1) refletem a importância relativa
    dessas informações na decisão final.

-   **Definindo a Taxa de Aprendizagem:** A `taxa_aprendizagem = 0.1`
    controla a velocidade com que o Perceptron ajusta seus pesos a cada
    erro. Imagine que a taxa de aprendizado é como a sensibilidade de
    uma balança.

    Uma taxa alta significa que o modelo fará grandes ajustes nos pesos
    a cada erro, o que pode acelerar o aprendizado, mas também torná-lo
    instável. Uma taxa baixa, por outro lado, fará com que o modelo faça
    pequenos ajustes, tornando o processo mais lento, porém mais
    preciso.

-   **Contabilizando os Erros:** `erros_totais = 0.0` funciona como um
    contador de erros. Durante o treinamento, o Perceptron compara suas
    previsões com os valores reais, e cada erro é contabilizado. Esse
    contador nos ajuda a avaliar o quão bem o modelo está aprendendo.

    Imagine um treinamento para o Perceptron identificar potenciais
    compradores em um marketplace. A cada vez que o modelo classifica
    incorretamente um usuário como comprador ou não comprador, o
    contador de erros é incrementado, fornecendo uma medida da precisão
    do modelo.

-   **Limitando as Épocas de Treinamento:** `limite_epocas = 20` define
    o número máximo de vezes que o Perceptron analisará todos os dados
    de treinamento. Essa é uma forma de controlar o tempo de treinamento
    do modelo.

    No caso de um Perceptron que aprende a prever o tempo de entrega de
    um produto com base na localização do vendedor e do comprador, tipo
    de frete e histórico de entregas. Cada época de treinamento seria
    como analisar todas as entregas já realizadas, permitindo que o
    modelo ajuste seus pesos para fazer previsões mais precisas.

-   **Fornecendo os Dados de Treinamento:** `dados_treinamento` é a base
    do aprendizado do Perceptron. É um conjunto de dados com informações
    sobre diferentes eventos, produtos ou usuários, juntamente com a
    classificação correta para cada um deles.

    Neste código, temos um conjunto de dados (x1, x2) e a indicação do
    resultado gerado (`saida_desejada = 1` ou `0`).

    Imagine que estamos treinando o Perceptron para identificar produtos
    com potencial de viralização. Os dados de treinamento conteriam
    informações sobre produtos que se tornaram virais no passado,
    juntamente com aqueles que não tiveram o mesmo sucesso.

## Treinamento do Perceptron

Com os parâmetros definidos e os dados de treinamento preparados, o
Perceptron inicia seu processo de aprendizado. Esse processo funciona
como um jogo de ajuste fino, onde o Perceptron tenta encontrar os
melhores valores para os pesos, de forma a fazer previsões cada vez mais
precisas.

In [2]:
# Inicia o treinamento do Perceptron
for epoca in range(1, limite_epocas + 1):
    print(f"\n Época: {epoca}\n")
    # Sinaliza se houve erro na época atual, inicialmente definido como False
    erro_na_epoca = False

    # Itera sobre cada dado de treinamento
    for dado in dados_treinamento:
        # Extrai os valores de entrada e o valor de referência do dado atual
        x1 = dado["x1"]
        x2 = dado["x2"]
        saida_desejada = dado["saida_desejada"]

        # Exibe os valores dos pesos atuais
        w1 = peso[0]
        w2 = peso[1]

        # Calcula a saída do perceptron (u)
        u = (x1 * peso[0]) + (x2 * peso[1])
        # Aplica a função de ativação (step function):
        # se u >= 0, valor_saida = 1, caso contrário, valor_saida = 0
        valor_saida = 1 if u >= 0 else 0

        # Calcula o erro: diferença entre o valor de referência e a saída do perceptron
        erro = saida_desejada - valor_saida

        # Acumula o erro total
        erros_totais += abs(erro)

        # Se houve erro na iteração atual, define erro_na_epoca como True
        if erro != 0:
            erro_na_epoca = True
            # Ajusta os pesos (w1 e w2) usando a regra delta
            peso[0] += taxa_aprendizagem * erro * x1
            peso[1] += taxa_aprendizagem * erro * x2

        # Mostra as informações da iteração atual (época, entradas, saída, pesos)
        print(
            f" X1: {x1:.1f}, X2: {x2:.1f} | W1: {w1:.2f}, W2: {w2:.2f} | Y: {
                saida_desejada}, "
            f"Y-predito: {valor_saida} | Erro: {
                erro}, Erro acumulado: {erros_totais}"
        )
        if erro_na_epoca:
            print(
                f" Ajuste de pesos: Novo W1 = {
                    peso[0]:.2f}, Novo W2 = {peso[1]:.2f}\n"
            )
        if not erro_na_epoca:
            print(" Ajuste de pesos: Não foi necessário.\n")
    print("-=-" * 31)

    # Verifica o critério de parada: se não houve erro na época (erro_na_epoca = False)
    if not erro_na_epoca:
        # Imprime mensagem de sucesso, indicando a época em que o erro zero foi atingido
        print(
            f"\n Modelo treinado com sucesso na época {
                epoca}! Erro zero atingido."
        )
        # Imprime os valores finais dos pesos (w1 e w2)
        print(
            f" Valores ideais de peso: W1 = {
                peso[0]:.2f}, W2 = {peso[1]:.2f}\n"
        )
        # Encerra o loop de treinamento
        break
# Caso contrário, se o loop atingir o limite de épocas sem sucesso
else:
    # Imprime mensagem informando que o limite de épocas foi atingido
    print(
        f"\n O limite de {
            limite_epocas} épocas foi atingido sem alcançar erro zero."
    )
    # Indica que o perceptron pode não ter convergido
    print(" O perceptron não pôde ser treinado com sucesso.\n")


 Época: 1

 X1: 0.5, X2: 0.8 | W1: 0.20, W2: -0.10 | Y: 1, Y-predito: 1 | Erro: 0, Erro acumulado: 0.0
 Ajuste de pesos: Não foi necessário.

 X1: 0.2, X2: 0.4 | W1: 0.20, W2: -0.10 | Y: 0, Y-predito: 1 | Erro: -1, Erro acumulado: 1.0
 Ajuste de pesos: Novo W1 = 0.18, Novo W2 = -0.14

-=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=-

 Época: 2

 X1: 0.5, X2: 0.8 | W1: 0.18, W2: -0.14 | Y: 1, Y-predito: 0 | Erro: 1, Erro acumulado: 2.0
 Ajuste de pesos: Novo W1 = 0.23, Novo W2 = -0.06

 X1: 0.2, X2: 0.4 | W1: 0.23, W2: -0.06 | Y: 0, Y-predito: 1 | Erro: -1, Erro acumulado: 3.0
 Ajuste de pesos: Novo W1 = 0.21, Novo W2 = -0.10

-=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=-

 Época: 3

 X1: 0.5, X2: 0.8 | W1: 0.21, W2: -0.10 | Y: 1, Y-predito: 1 | Erro: 0, Erro acumulado: 3.0
 Ajuste de pesos: Não foi necessário.

 X1: 0.2, X2: 0.4 | W1: 0.21, W2: -0.10 | Y: 0, Y-predito: 1 | Erro: -1, Erro acumu

### Explicação do Código:

O código acima representa o cerne do processo de treinamento do
Perceptron, um ciclo iterativo de aprendizado que se repete a cada
época, refinando os pesos do modelo para minimizar os erros de previsão.
Vamos analisar cada etapa em detalhes:

1.  **Controlando as Épocas:** O primeiro laço `for` controla o número
    de épocas de treinamento, onde cada época funciona como um ciclo
    completo de aprendizado, no qual o Perceptron analisa todos os dados
    de treinamento, ajusta seus pesos e calcula o erro total. O limite
    de épocas, definido anteriormente como 20, determina o número máximo
    desses ciclos.

2.  **Aplicando os Dados de Treinamento:** Dentro de cada época, o
    segundo laço `for` percorre cada dado de treinamento
    individualmente. É como se o Perceptron estivesse examinando cada
    exemplo de produto, um por um, para aprender com suas
    características (x1, x2) e sua classificação correta
    (`saida_desejada`).

3.  **Extraindo as Informações:** Para cada produto nos dados de
    treinamento, o código extrai suas características (x1, x2) e a
    classificação correta (`saida_desejada`). Essas informações serão
    usadas para calcular a saída do Perceptron e, posteriormente,
    ajustar os pesos.

4.  **Armazenando os Pesos Atuais:** As variáveis `w1` e `w2` armazenam
    os valores atuais dos pesos. Isso é feito para facilitar a
    visualização dos pesos durante o processo de treinamento.

5.  **Calculando a Saída do Perceptron:** Aqui é aplicada a fórmula
    matemática do Perceptron. A “soma ponderada” (u) é calculada
    multiplicando as características do produto (x1, x2) pelos seus
    respectivos pesos (w1, w2) e somando os resultados. No caso do
    exemplo da tentativa de prever se um produto irá viralizar, essa
    soma ponderada representa a “avaliação” do Perceptron sobre o
    produto, indicando sua propensão a ser um sucesso de vendas ou não.

6.  **Aplicando a Função de Ativação:** A “soma ponderada” (u) é então
    passada pela função de ativação “degrau”. Essa função define a saída
    do Perceptron (`valor_saida`) com base na “soma ponderada”: se u for
    maior ou igual a zero, a saída é 1 (sucesso de vendas); caso
    contrário, a saída é 0 (não sucesso).

7.  **Calculando o Erro:** O Perceptron compara sua previsão
    (`valor_saida`) com a classificação correta (`saida_desejada`). A
    diferença entre esses valores representa o “erro” do modelo. É a
    partir do erro que o Perceptron ajusta seus pesos, buscando
    minimizar a diferença entre suas previsões e os valores reais.

8.  **Ajustando os Pesos:** Se houver erro na previsão, os pesos (w1,
    w2) são ajustados usando a regra delta. Essa regra utiliza a taxa de
    aprendizagem para controlar o tamanho do ajuste. O objetivo do
    ajuste é fazer com que o Perceptron se aproxime da classificação
    correta a cada iteração.

9.  **Mostrando as Informações da Época:** O código exibe as informações
    da época atual, incluindo as entradas e os pesos usados na
    interação, a saída desejada, a saída predita, o erro, o erro
    acumulado e o ajuste de pesos. Essa visualização detalhada permite
    acompanhar o progresso do treinamento e entender como os pesos são
    ajustados a cada época.

10. **Verificando o Critério de Parada:** Ao final de cada época, o
    código verifica se o Perceptron atingiu o erro zero, ou seja, se ele
    classificou corretamente todos os dados de treinamento. Se o erro
    zero for atingido, o treinamento é interrompido com sucesso, e os
    pesos finais do modelo são exibidos. Caso contrário, o treinamento
    continua até que o limite de épocas seja atingido.

O processo de treinamento do Perceptron busca encontrar um ponto de
equilíbrio, ajustando os pesos de forma que o erro seja minimizado. Esse
processo de aprendizado, embora relativamente simples, é a base para o
desenvolvimento de modelos mais complexos e poderosos de inteligência
artificial.

## Interpretação dos Resultados

A análise do resultado gerado ao rodar o modelo nos permite avaliar o
processo de aprendizado do Perceptron e determinar se o treinamento foi
bem-sucedido. O sucesso do treinamento é definido pela capacidade do
modelo de classificar corretamente os dados de treinamento, o que, nesse
caso, significa atingir o erro zero.

Observando os resultados, é possível identificar que o Perceptron
atingiu o erro zero na época 12. Isso significa que o modelo precisou
passar 12 vezes pelos dados de treinamento para encontrar um conjunto de
pesos que tivesse capacidade de classificar os dados corretamente.

A necessidade de várias épocas para alcançar o erro zero significa que,
apesar dos dados de treinamento poderem ser separados em duas classes
por uma linha (linearmente separáveis), essa separação não era
imediatamente óbvia para o Perceptron com os pesos aleatórios que ele
recebeu no início do treinamento. Sendo assim, os pesos iniciais não
eram adequados para classificar os dados corretamente, exigindo um
processo de ajuste gradual ao longo das épocas.

Em cada uma dessas épocas, o Perceptron ajustou seus pesos em resposta
aos erros de classificação, buscando minimizar a distância entre suas
previsões e os valores reais. A cada ajuste, o modelo se aproximava da
solução ideal, até que, na época 12, todos os dados foram classificados
corretamente.

Os pesos finais encontrados, `W1 = 0.23` e `W2 = -0.14`, refletem o
conhecimento adquirido pelo modelo durante o treinamento. A análise
desses pesos nos permite entender como o Perceptron aprendeu a
classificar os dados:

-   **W1 positivo (0.23):** Indica que a característica x1 tem uma
    influência positiva na decisão final do modelo. Um valor maior para
    x1 aumenta a propensão do Perceptron a gerar uma saída “1”, que,
    dependendo do problema em questão, pode representar a recomendação
    de um produto, a classificação de um cliente como potencial
    comprador, ou qualquer outro evento que se deseje prever.

-   **W2 negativo (-0.14):** Indica que a característica x2 tem uma
    influência negativa na decisão final. Diferente de x1, um valor
    maior para x2 diminui a propensão do modelo a gerar uma saída “1”,
    atuando como um contraponto à influência de x1.

Essa análise dos pesos demonstra a capacidade do Perceptron de ponderar
diferentes características e aprender relações complexas entre os dados.

É fundamental destacar que, se o limite de épocas fosse definido como 3,
o treinamento seria interrompido antes do Perceptron atingir o erro
zero. Com apenas 3 épocas, o modelo não teria tempo suficiente para
ajustar seus pesos de forma a classificar corretamente todos os dados de
treinamento. Os pesos finais, nesse cenário, representariam uma solução
subótima, comprometendo a capacidade de generalização do modelo, ou
seja, sua capacidade de ter um bom desempenho em dados não vistos
durante o treinamento.

## Conclusão

O Perceptron, mesmo em sua forma mais básica, demonstra o potencial do
aprendizado de máquina para solucionar problemas complexos em uma
variedade de setores. Suas aplicações, que vão desde a segmentação
precisa de clientes e a previsão acurada da demanda até a recomendação
personalizada de produtos e a otimização da experiência de busca,
ilustram sua capacidade de extrair padrões intrincados de dados e gerar
previsões robustas. Com isso, o Perceptron abre um leque de
oportunidades para aprimorar a eficiência operacional, personalizar a
interação com o cliente, impulsionar as vendas e, sendo assim, auxiliar
na construção de um ecossistema de negócios mais inteligente, ágil e
responsivo às atuais demandas do mercado.